src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs

```diff
@@ -20,36 +20,40 @@ namespace Castle.Windsor.Extensions.DependencyInjection.Scope
 	using Castle.Core;
 	using Castle.MicroKernel;
 	using Castle.MicroKernel.Lifestyle.Scoped;
-	
+
 	internal class ExtensionContainerScope : ILifetimeScope, IDisposable
 	{
 		public static ExtensionContainerScope Current => current.Value;
 		public static string TransientMarker = "Transient";
 		protected static readonly AsyncLocal<ExtensionContainerScope> current = new AsyncLocal<ExtensionContainerScope>();
 		private readonly ExtensionContainerScope parent;
+		private readonly ExtensionContainerRootScope rootScope;
 		private readonly IScopeCache scopeCache;
 
-		protected ExtensionContainerScope(ExtensionContainerScope parent)
+		protected ExtensionContainerScope(
+			ExtensionContainerScope parent,
+			ExtensionContainerRootScope rootScope)
 		{
 			scopeCache = new ScopeCache();
-			if(parent == null)
-			{
-				this.parent = ExtensionContainerRootScope.RootScope;
-			}
-			else
-			{
-				this.parent = parent;
-			}
+			this.parent = parent ?? rootScope;
+			this.rootScope = rootScope;
 		}
 
-		public static ExtensionContainerScope BeginScope(ExtensionContainerScope parent)
+		public ExtensionContainerRootScope RootScope
+			=> this as ExtensionContainerRootScope ?? rootScope;
+
+		public static ExtensionContainerScope BeginScope()
 		{
-			var scope = new ExtensionContainerScope(parent);
+			var parent = Current;
+			var rootScope = Current?.RootScope; 
+			if (rootScope == null)
+				throw new ArgumentNullException(nameof(rootScope));
```

Comment:
`ArgumentNullException` for parameter `rootScope` doesn't make sense anymore since it isn't a parameter.

Reply:
Root scope, when creating a scope must not be null. 
This couls happen if you ty to BeginScope, before "BeginRootScope" was called. I.e. if the user has there own WindsorServiceProviderFactory

Reply:
Improved